spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteDataFilesSparkAction.java

```diff
@@ -0,0 +1,321 @@
+          return Pair.of(new FileGroupInfo(UUID.randomUUID().toString(), myJobIndex, myPartIndex, e.getKey()), tasks);
+        })
+    );
+
+    int maxConcurrentFileGroupActions = PropertyUtil.propertyAsInt(options(),
+        RewriteDataFiles.MAX_CONCURRENT_FILE_GROUP_ACTIONS,
+        RewriteDataFiles.MAX_CONCURRENT_FILE_GROUP_ACTIONS_DEFAULT);
+
+    int maxCommits = PropertyUtil.propertyAsInt(options(),
+        RewriteDataFiles.PARTIAL_PROGRESS_MAX_COMMITS,
+        RewriteDataFiles.PARTIAL_PROGRESS_MAX_COMMITS_DEFAULT);
+
+    boolean partialProgressEnabled = PropertyUtil.propertyAsBoolean(options(),
+        RewriteDataFiles.PARTIAL_PROGRESS_ENABLED,
+        RewriteDataFiles.PARTIAL_PROGRESS_ENABLED_DEFAULT);
+
+    int groupsPerCommit = partialProgressEnabled ? totalGroups / maxCommits : totalGroups;
+
+    AtomicBoolean stillRewriting = new AtomicBoolean(true);
+    ConcurrentLinkedQueue<String> completedRewrite = new ConcurrentLinkedQueue<>();
+    ConcurrentHashMap<String, Pair<FileGroupInfo, FileGroupRewriteResult>> results = new ConcurrentHashMap<>();
+    ConcurrentLinkedQueue<String> completedCommits = new ConcurrentLinkedQueue<>();
+
+    ExecutorService rewriteService = Executors.newFixedThreadPool(maxConcurrentFileGroupActions,
+        new ThreadFactoryBuilder().setNameFormat("Rewrite-Service-%d").build());
+
+    ExecutorService commiterService = Executors.newSingleThreadExecutor(
+        new ThreadFactoryBuilder().setNameFormat("Committer-Service").build());
+
+    commiterService.execute(() -> {
+      try {
+        while (stillRewriting.get() || completedRewrite.size() > groupsPerCommit) {
+          Thread.yield();
+          if (completedRewrite.size() > groupsPerCommit) {
+            // Gather a group of rewrites to do a commit
+            Set<String> batch = Sets.newHashSetWithExpectedSize(groupsPerCommit);
+            for (int i = 0; i < groupsPerCommit; i++) {
+              batch.add(completedRewrite.poll());
+            }
+            // Running commit on the group we gathered
+            try {
+              commitFileGroups(batch);
+            } catch (Exception e) {
+              if (!partialProgressEnabled) {
+                LOG.error("Failure during rewrite commit process, partial progress not enabled. Rethrowing", e);
+                throw e;
+              } else {
+                LOG.error("Failure during rewrite commit process, partial progress enabled. Ignoring", e);
+              }
+            }
+            completedCommits.addAll(batch);
+          }
+        }
+        Set<String> finalBatch = Sets.newHashSetWithExpectedSize(groupsPerCommit);
+        while (completedRewrite.size() != 0) {
+          finalBatch.add(completedRewrite.poll());
+        }
+        if (finalBatch.size() != 0) {
+          // Final Commit
+          commitFileGroups(finalBatch);
+          completedCommits.addAll(finalBatch);
+        }
+      } catch (Exception e) {
+        if (!partialProgressEnabled) {
```

Comment:
Or we actually assume this only happens if all rewrites have completed?

Reply:
This is only in the case that partialProgress is not enabled, in that case we either have 1 commit (this section of the code) so there is no need to stop rewrites since they have all finished